lib/value/src/kind/collection/exact.rs

```diff
@@ -0,0 +1,349 @@
+use std::{
+    collections::BTreeMap,
+    ops::{BitAnd, BitOr, BitXor},
+};
+
+use crate::Kind;
+
+/// The exact type-state of an [`Unknown`](super::Unknown) value in a collection.
+///
+/// This is its own type, to avoid infinite recursion for nested collection [`Kind`]s.
+#[derive(Debug, Clone, Copy, Eq, PartialEq)]
+#[allow(clippy::struct_excessive_bools)]
+pub struct Exact {
+    pub(super) bytes: bool,
+    pub(super) integer: bool,
+    pub(super) float: bool,
+    pub(super) boolean: bool,
+    pub(super) timestamp: bool,
+    pub(super) regex: bool,
+    pub(super) null: bool,
+
+    // We don't need to support nested objects, because `Exact` is only used for "unknown" fields.
+    // So it only applies to a single level. If we wanted to set a nested field, we would have
+    // a "known" field (e.g. `foo`) and then have that contain either `Unknown::Any` or
+    // `Unknown::Exact`.
+    pub(super) object: bool,
+    pub(super) array: bool,
+}
```

Comment:
I explicitly didn't want to use it because we use(d) it in the old implementation, and while it is cool and all at first, it just comes with an API that is broader than we need, and it caused more confusion than it was worth over the last year+.

The most frequently confusing example I can remember is that bitflags has an `all()` function to initialize the type, but we call it `any()`, as that better fits the mental model of progressive type checking, but every so often, that method popped up in the code because people's editors auto-completed it, and it _works_, but it's confusing to see it being used next to `any()`.

TL;DR: at a surface, it makes sense, but after having used it for more than a year, I found it to be more confusing than worthwhile.

Reply:
I'll add that we _could_ wrap the `bitflags` type in a newtype, but ultimately, I just don't think the juice is worth the squeeze.